File: LinkedList.cpp
#include <iostream>
using namespace std;

class Node
{
public:
	int data;
	Node *next;
};

void printList(Node *head)
{
	while (head != NULL)
	{
		cout << head->data << " ";
		head = head->next;
	}
	cout << endl;
}

void insertAtEnd(Node *head, int data)
{
	while (head)
	{
		if (head->next == NULL)
		{
			Node *newNode = new Node();
			newNode->data = data;
			newNode->next = NULL;
			head->next = newNode;
			break;
		}
		head = head->next;
	}
}

void insertAtStart(Node **head, int data)
{
	Node *newNode = new Node();
	newNode->data = data;
	newNode->next = *head;
	*head = newNode;
}

void insertAtMid(Node **head, int data, int n)
{
	Node *temp = new Node();
	temp = *head;
	while (temp && n--)
	{
		temp = temp->next;
	}
	Node *newNode = new Node();
	newNode->next = temp->next;
	newNode->data = data;
	temp->next = newNode;
}

int main()
{
	Node *head = new Node();
	head = NULL;

	insertAtStart(&head, 1);
	printList(head);

	insertAtEnd(head, 2);
	insertAtEnd(head, 3);
	printList(head);

	insertAtEnd(head, 5);
	printList(head);

	insertAtStart(&head, 0);
	printList(head);

	insertAtMid(&head, 11, 3);
	printList(head);
}

File: js/linkedList.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
    this.size = 0;
  }

  add(element) {
    var node = new Node(element);
    var current;

    if (this.head == null) {
      this.head = node;
    } else {
      current = this.head;
      while (current.next) {
        current = current.next;
      }
      current.next = node;
    }
    this.size++;
  }

  insertAt(element, index) {
    if (index < 0 || index > this.size) {
      return console.log("Invalid Index");
    }
    var node = new Node(element);
    var current, previous;

    current = this.head;
    if (index == 0) {
      node.next = this.head;
      this.head = node;
    } else {
      current = this.head;
      var i = 0;
      while (i < index) {
        i++;
        previous = current;
        current = current.next;
      }
      node.next = current;
      previous.next = node;
    }
    this.size++;
  }

  removeFrom(index) {
    if (index < 0 || index > this.size) {
      return console.log("Invalid Index");
    }
    var current,
      previous,
      i = 0;
    current = this.head;
    previous = current;

    if (index == 0) {
      this.head = current.next;
    } else {
      while (i < index) {
        i++;
        previous = current;
        current = current.next;
      }
      previous.next = current.next;
    }
    this.size--;
    return current.data;
  }

  remove(element) {
    var current = this.head;
    var previous = null;

    while (current != null) {
      if (current.data == element) {
        if (previous == null) {
          this.head = current.next;
        } else {
          previous.next = current.next;
        }
        this.size--;
        return current.data;
      }
      previous = current;
      current = current.next;
    }
    return -1;
  }

  search(element) {
    var current = this.head;
    var i = 0;
    while (current) {
      if (current.data === element) {
        return i;
      }
      current = current.next;
      i++;
    }
    return -1;
  }

  printList() {
    var current = this.head;
    var str = "";
    while (current) {
      str += current.data + " ";
      current = current.next;
    }
    console.log(str);
  }
}

var ll = new LinkedList();
ll.add(10);
ll.add(20);
ll.add(30);
ll.add(40);
ll.add(50);
console.log("Removed: ", ll.removeFrom(1));
console.log("Removed: ", ll.remove(30));
console.log("Found at: ", ll.search(20));
console.log("Found at: ", ll.search(10));
ll.printList();
